File: week1/w1d2/iteration.rb
def find_number
  i = 251
  loop do
    return i if i % 7 == 0
    i += 1
  end
end

# p find_number

def factors(number)
  1.upto(number).select { |factor| number % factor == 0 }
end

# p factors(945)

def bubble_sort(array)
  sorted = false
  until sorted
    sorted = true
    
    0.upto(array.size - 2).each do |i|
      if array[i] > array[i + 1]
        array[i], array[i + 1] = array[i + 1], array[i]
        sorted = false
      end
    end
  end
  
  array
end

# p bubble_sort([2,1,5,3,8])

def substrings(string)
  substrings = []
  
  (0...string.length).each do |start_index|
    (start_index...string.length).each do |end_index|
      substrings << string[start_index..end_index]
    end
  end
  
  substrings
end

# p substrings('cat')

def subwords(string)
  dictionary = File.read('dictionary.txt').split(/\n/)
  substrings(string).select { |substring| dictionary.include? (substring) }
end

# p subwords('cat')

File: week5/w5d4/iteration.js
var bubbleSort = function(array){
  var sorted = false;
  while (!sorted) {
    sorted = true;
    for (var i = 0; i < array.length - 1; i++){
      if (array[i] > array[i+1]){
        var tmp = array[i];
        array[i] = array[i+1];
        array[i+1] = tmp;
        sorted = false;
      }
    }
  }
  return array;
};

// console.log(bubbleSort([1,7,4,5,2,1]));

var substrings = function(string){
  var substringsArray = [];
  for (var start_index = 0; start_index < string.length; start_index++){
    for (var end_index = start_index; end_index < string.length; end_index++){
      substringsArray.push(string.slice(start_index, end_index+1));
    }
  }
  return substringsArray;
}

// console.log(substrings('reza'));

String.prototype.substrings = function(){
  var substringsArray = [];
  for (var start_index = 0; start_index < this.length; start_index++){
    for (var end_index = start_index; end_index < this.length; end_index++){
      substringsArray.push(this.slice(start_index, end_index+1));
    }
  }
  return substringsArray;
}


// console.log("random string".substrings());

Array.prototype.bubbleSort = function(){
  var sorted = false;
  while (!sorted) {
    sorted = true;
    for (var i = 0; i < this.length - 1; i++){
      if (this[i] > this[i+1]){
        var tmp = this[i];
        this[i] = this[i+1];
        this[i+1] = tmp;
        sorted = false;
      }
    }
  }
  return this;
};

console.log([2,3,6,0,-11,293,2].bubbleSort());
